src/Avalonia.Controls/MaskedTextBox.cs

```diff
@@ -0,0 +1,394 @@
+        /// </summary>
+        public new char PasswordChar
+        {
+            get => GetValue(PasswordCharProperty);
+            set => SetValue(PasswordCharProperty, value);
+        }
+
+        /// <summary>
+        /// Gets or sets the character used to represent the absence of user input in MaskedTextBox.
+        /// </summary>
+        public char PromptChar
+        {
+            get => GetValue(PromptCharProperty);
+            set => SetValue(PromptCharProperty, value);
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating if selected characters should be reset when the prompt character is pressed.
+        /// </summary>
+        public bool ResetOnPrompt
+        {
+            get => _resetOnPrompt;
+            set
+            {
+                SetAndRaise(ResetOnPromptProperty, ref _resetOnPrompt, value);
+                if (MaskProvider != null)
+                {
+                    MaskProvider.ResetOnPrompt = value;
+                }
+
+            }
+        }
+        /// <summary>
+        /// Gets or sets a value indicating if selected characters should be reset when the space character is pressed.
+        /// </summary>
+        public bool ResetOnSpace
+        {
+            get => _resetOnSpace;
+            set
+            {
+                SetAndRaise(ResetOnSpaceProperty, ref _resetOnSpace, value);
+                if (MaskProvider != null)
+                {
+                    MaskProvider.ResetOnSpace = value;
+                }
+
+            }
+
+
+        }
+        Type IStyleable.StyleKey => typeof(TextBox);
+
+        protected override void OnGotFocus(GotFocusEventArgs e)
+        {
+            if (HidePromptOnLeave == true && MaskProvider != null)
+            {
+                Text = MaskProvider.ToDisplayString();
+            }
+            base.OnGotFocus(e);
+        }
+
+        protected override async void OnKeyDown(KeyEventArgs e)
+        {
+            if (MaskProvider is null)
+            {
+                base.OnKeyDown(e);
+                return;
+            }
+            var keymap = AvaloniaLocator.Current.GetService<PlatformHotkeyConfiguration>();
+
+            bool Match(List<KeyGesture> gestures) => gestures.Any(g => g.Matches(e));
+
+            if (Match(keymap.Paste))
+            {
+                var text = await ((IClipboard)AvaloniaLocator.Current.GetService(typeof(IClipboard))).GetTextAsync();
+
+                if (text is null)
+                    return;
+
+                foreach (var item in text)
+                {
+                    var index = GetNextCharacterPosition(CaretIndex);
+                    if (MaskProvider.InsertAt(item, index))
+                    {
+                        CaretIndex = ++index;
+                    }
+                }
+
+                Text = MaskProvider.ToDisplayString();
+                e.Handled = true;
```

Comment:
Here it would be nice to handle undo\redo also, but I am not sure how to do that since undoRedoHelper is private.

Reply:
You can make https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/TextBox.cs#L1337 protected